allennlp/tests/training/event_based_trainer_test.py

```diff
@@ -0,0 +1,131 @@
+# pylint: disable=invalid-name,too-many-public-methods,protected-access
+from typing import Iterable
+
+import torch
+
+from allennlp.common.params import Params
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.common.tqdm import Tqdm
+from allennlp.data import Vocabulary
+from allennlp.data.dataset_readers import SequenceTaggingDatasetReader
+from allennlp.data.iterators import BasicIterator
+from allennlp.data.iterators.data_iterator import TensorDict
+from allennlp.models.simple_tagger import SimpleTagger
+from allennlp.training import Trainer
+from allennlp.training.event_based_trainer.callback import Callback
+from allennlp.training.event_based_trainer.event_based_trainer import EventBasedTrainer
+from allennlp.training.event_based_trainer.events import Events
+
+
+class TrainOneBatch(Callback):
+    """
+    Trains one batch by running
+        optimizer.zero_grad -> model.forward -> loss.backward -> optimizer.step
+    """
+    def __init__(self, optimizer):
+        self.optimizer = optimizer
+
+    def call(self, trainer: EventBasedTrainer) -> None:
+        self.optimizer.zero_grad()
```

Comment:
Pretty sure this state is going to have to live in the trainer, not here.  Otherwise the `from_params` stuff doesn't work out very well.  It's possible that the `extras` stuff just works and passes it to here just fine, but then you'd still break existing config files if you want to replace our existing trainer with this.

Reply:
I don't want to replace our existing trainer with this, I want to replace it with a subclass of this (that has a constructor that looks like the current trainer's constructor), and that subclass instance would own the optimizer. My thinking is that maybe in some other trainer variant I might have multiple optimizers, or something weird like that.

Reply:
Hmm, yeah, I can see that.  If you require the `Trainer` to own all of that state, you make it hard to implement callbacks that require other kinds of state.  That's a really good point.  You're right here, and I retract what I said.  We just need to be sure that passing arguments through to the callbacks works as it should.